### PR TITLE
Add data about job posting interactions to hiring staff dashboard

### PR DESCRIPTION
**Pull request by pezholio**
## Trello card URL:

https://trello.com/c/j07qRe5U/651-add-data-about-job-posting-interactions-to-hiring-staff-dashboard-to-highlight-job-posting-activity

## Changes in this PR:

This adds the page views statistics that are already being collected to the hiring staff dashboard, as well as starting to gather the stats for the 'get more information' clicks in the same way (append a Redis counter, then update them in the database on a daily basis). There's also a one off rake task and task definition to take the stats that we've gathered on the 'get more information' clicks using the PublicActivity gem, so we can backfill the clicks for older vacancies.

### Before

image

### After

image

## Next steps:

- [x] Terraform deployment required?

### CODE REVIEW

File: app/controllers/vacancies_controller.rb [hunk 1]
```diff
@@ -30,7 +30,8 @@ def show
     return redirect_to(job_path(vacancy), status: :moved_permanently) if old_vacancy_path?(vacancy)
 
     @vacancy = VacancyPresenter.new(vacancy)
-    TrackVacancyPageViewJob.perform_later(vacancy.id) unless authenticated?
+
+    VacancyPageView.new(vacancy).track unless authenticated?
```
Comment (robbiepaul) [inline app/controllers/vacancies_controller.rb]: I like this, like you said it's just swapping one Redis call for another makes sense to cut out the middleman

File: app/jobs/persist_vacancy_get_more_info_click_job.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class PersistVacancyGetMoreInfoClickJob < ApplicationJob
+  queue_as :page_view_collector
```
Comment (robbiepaul) [inline app/jobs/persist_vacancy_get_more_info_click_job.rb]: Should we change this queue name to something like `persist_counters` or something similar/generic?

File: spec/presenters/school_vacancy_presenter_spec.rb [hunk 3]
```diff
@@ -22,4 +22,24 @@
       end
     end
   end
+
+  describe 'get_more_info_clicks' do
+    let(:vacancy) { create(:vacancy, total_get_more_info_clicks: get_more_info_clicks) }
+
+    context 'when page views are present' do
```
Comment (robbiepaul) [inline spec/presenters/school_vacancy_presenter_spec.rb]: I think this should be `when get more info clicks are present`

File: spec/presenters/school_vacancy_presenter_spec.rb [hunk 4]
```diff
@@ -22,4 +22,24 @@
...
+      let(:get_more_info_clicks) { 100 }
+
+      it 'returns the number' do
+        expect(presenter.get_more_info_clicks).to eq(get_more_info_clicks)
+      end
+    end
+
+    context 'when page views are not present' do
```
Comment (robbiepaul) [inline spec/presenters/school_vacancy_presenter_spec.rb]: same as above

File: lib/tasks/vacancies_statistics_backfill_info_clicks.rake [hunk 5]
```diff
@@ -0,0 +1,18 @@
+namespace :vacancies do
+  desc 'Refreshes the cached pageviews for listed non expired job vacancies'
+  namespace :statistics do
+    namespace :backfill do
+      task info_clicks: :environment do
+        Vacancy.where(total_get_more_info_clicks: nil).each do |vacancy|
```
Comment (robbiepaul) [inline lib/tasks/vacancies_statistics_backfill_info_clicks.rake]: 👍

File: app/jobs/persist_vacancy_get_more_info_click_job.rb (hunk 2 above)
Comment (pezholio) [reply app/jobs/persist_vacancy_get_more_info_click_job.rb]: Yep, I thought that at the time, but forgot to do it!

**Comment by pezholio** [comment]
@robbiepaul Ready for another looksee now 👍

File: app/controllers/interests_controller.rb [hunk 6]
```diff
@@ -1,6 +1,7 @@
 class InterestsController < ApplicationController
   def new
     vacancy = Vacancy.find(vacancy_id)
+    VacancyGetMoreInfoClick.new(vacancy).track
```
Comment (robbiepaul) [inline app/controllers/interests_controller.rb]: It's not on the card but do you think we should be consistent with the page view counter and filter out authenticated users who click 'get more info'?

Comment (pezholio) [reply app/controllers/interests_controller.rb]: Yeah, good call

**Comment by tahb** [comment]
[Add data about job posting interactions to hiring staff dashboard to highlight job posting activity](https://trello.com/c/j07qRe5U/651-add-data-about-job-posting-interactions-to-hiring-staff-dashboard-to-highlight-job-posting-activity)

**Pull request merged**